core/txpool/subpool.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package txpool
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto/kzg4844"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+// Transaction is a helper struct to group together a canonical transaction with
+// satellite data items that are needed by the pool but are not part of the chain.
+type Transaction struct {
+	Tx *types.Transaction // Canonical transacrion
+
+	BlobTxBlobs   []kzg4844.Blob       // Blobs needed by the blob pool
+	BlobTxCommits []kzg4844.Commitment // Commitments needed by the blob pool
+	BlobTxProofs  []kzg4844.Proof      // Proofs needed by the blob pool
+}
+
+// SubPool represents a specialized transaction pool that lives on its own (e.g.
+// blob pool). Since independent of how many specialized pools we have, they do
+// need to be updated in lockstep and assemble into one coherent view for block
+// production, this interface defines the common methods that allow the primary
+// transaction pool to manage the subpools.
+type SubPool interface {
+	// Filter is a selector used to decide whether a transaction whould be added
+	// to this particular subpool.
+	Filter(tx *types.Transaction) bool
+
+	// Init sets the base parameters of the subpool, allowing it to load any saved
+	// transactions from disk and also permitting internal maintenance routines to
+	// start up.
+	//
+	// These should not be passed as a constructor argument - nor should the pools
+	// start by themselves - in order to keep multiple subpools in lockstep with
+	// one another.
+	Init(gasTip *big.Int, head *types.Header) error
+
+	// Close terminates any background processing threads and releases any held
+	// resources.
+	Close() error
```

Comment:
legacy pool doesn't ever return an error (only `nil`). You might skip this returnvalue and simplify the txpoo.go Close method err-handling a bit (unless you foresee the blobpool to have a reason to return errors on closing)

Reply:
Myeah, it's just a PITA to add errors after the fact everywhere, so might as well do it properly. Also, the blobpool does return an error since it has a database that can fail (2 actually).